src/stdlib_stats_distribution_normal.fypp

```diff
@@ -0,0 +1,332 @@
+
+        tr = rvs_norm_r${k1}$(loc % re, scale % re)
+        ti = rvs_norm_r${k1}$(loc % im, scale % im)
+        res = cmplx(tr, ti, kind=${k1}$)
+    end function rvs_norm_${t1[0]}$${k1}$
+
+    #:endfor
+
+
+
+    #:for k1, t1 in REAL_KINDS_TYPES
+    function rvs_norm_array_${t1[0]}$${k1}$(loc, scale, array_size) result(res)
+        ${t1}$, intent(in) :: loc, scale
+        integer, intent(in) :: array_size
+        ${t1}$ :: res(array_size)
+        ${t1}$, parameter  ::  r = 3.442619855899_${k1}$, rr = 1.0_${k1}$ / r
+        ${t1}$ ::  x, y, re
+        integer :: hz, iz, i
+
+        if(scale == 0._${k1}$) call error_stop("Error(rvs_norm_array): Normal" &
+            //"distribution scale parameter must be non-zero")
+        if(.not. zig_norm_initialized) call zigset
+        do i = 1, array_size
+            iz = 0
+            hz = dist_rand(1_int32)
+            iz = iand( hz, 127 )
+            if( abs( hz ) < kn(iz) ) then
+                re = hz * wn(iz)
+            else
+                L1: do
+                    L2: if( iz == 0 ) then
+                        do
+                            x = -log( uni(1.0_${k1}$) ) * rr
+                            y = -log( uni(1.0_${k1}$) )
+                            if( y + y >= x * x ) exit
+                        end do
+                        re = r + x
+                        if( hz <= 0 ) re = -re
+                        exit L1
+                    end if L2
+                    x = hz * wn(iz)
+                    if( fn(iz) + uni(1.0_${k1}$) * (fn(iz-1) - fn(iz)) <        &
+                        exp(-HALF * x * x) ) then
+                        re = x
+                        exit L1
+                    end if
+
+                    hz = dist_rand(1_int32)
+                    iz = iand( hz, 127 )
+                    if( abs( hz ) < kn(iz) ) then
+                        re = hz * wn(iz)
+                        exit L1
+                    end if
+                end do L1
+            end if
+            res(i) = re * scale + loc
+        end do
+    end function rvs_norm_array_${t1[0]}$${k1}$
+
+    #:endfor
+
+
+
+    #:for k1, t1 in CMPLX_KINDS_TYPES
+    function rvs_norm_array_${t1[0]}$${k1}$(loc, scale, array_size) result(res)
+        ${t1}$, intent(in) :: loc, scale
+        integer, intent(in) :: array_size
+        integer :: i
+        ${t1}$ :: res(array_size)
+        real(${k1}$) :: tr, ti
+
+        do i = 1, array_size
+            tr = rvs_norm_r${k1}$(loc % re, scale % re)
+            ti = rvs_norm_r${k1}$(loc % im, scale % im)
+            res(i) = cmplx(tr, ti, kind=${k1}$)
+        end do
+    end function rvs_norm_array_${t1[0]}$${k1}$
+
+    #:endfor
+
+
+
+    #:for k1, t1 in REAL_KINDS_TYPES
+    impure elemental function pdf_norm_${t1[0]}$${k1}$(x, loc, scale) result(res)
+    !
+    ! Normal distribution probability density function
+    !
+        ${t1}$, intent(in) :: x, loc, scale
+        real :: res
```

Comment:
Here I think `res` should be of type ${t1}$ ? I noticed differences between this function and the results from Rs `dnorm`, when usign double-precision in Fortran, that are likely due to this.

Reply:
Note: The same kind of bug also occurs in the equivalent `uniform` module

Reply:
Just confirming that when I make the obvious fix in my local branch of stdlib, the results agree with R to within `5e-016`, when checking 401 evenly spaced points from -10 to 10 standard deviations in double precision. Whereas the differences reach around `1e-08` otherwise.

Reply:
It would be good if the unit tests were able to catch such precision issues (easy mistake to make or introduce in future when editing).

Reply:
May be due to my limited probability knowledge, I purposely restrict the results of all probability function value to be single precision. To me, the **probability** of 0.1234567 (single precision) would be the same as **probability** of 0.12345678901234567 (double precision) practically, thought mathematically there are different. May be I am wrong.

Reply:
Should we provide all precision for probability?

Reply:
Yes, we should provide all precision for all probability. For these kids of functions, the need for precision is similar to other branches of applied math. 

While in some cases single-precision is fine, we also do things that need higher precision (e.g. suppose these functions appear in an expression that is then numerically differentiated).

I think the solution is to make `res` be of kind `real(${k1}$)` in all of the functions. Casual tests of `pdf_uniform` and `cdf_uniform` with this approach gave results very close to R's `dnorm` and `pnorm` in double precision.

Reply:
Based on tests I ran just now, it should be: `real(${k1}$) :: res`

Reply:
So here we got a problem. The normal distribution is based on Marsaglia's fast ziggurat method which use double precision for the ziggurat table. It would be OK for the single and double precision probability but quardruple precision probability. For quardruple precision probability, any number beyond 16 significant figures will be meaningless.

Reply:
I see, thanks. In the first instance, I think it's OK if the random number generator does not give quad-precision accuracy - we can worry about that later if it's important. But for the pdf/cdf functions, we should be able to get full precision without any problems.

Reply:
But please add a statement in the specs about the precision of the random numbers being fundamentally limited to double precision by the algorithm (even if results are provided for quad precision).